packages/vite/src/node/utils.ts

```diff
@@ -320,12 +320,15 @@ export function injectQuery(url: string, queryToInject: string): string {
     url.replace(replacePercentageRE, '%25'),
     'relative:///',
   )
-  const { search, hash } = resolvedUrl
+  const { search, searchParams, hash } = resolvedUrl
   let pathname = cleanUrl(url)
   pathname = isWindows ? slash(pathname) : pathname
-  return `${pathname}?${queryToInject}${search ? `&` + search.slice(1) : ''}${
-    hash ?? ''
-  }`
+
+  return searchParams.get(queryToInject) === ''
+    ? `${pathname}${search}${hash ?? ''}`
+    : `${pathname}?${queryToInject}${search ? `&` + search.slice(1) : ''}${
+        hash ?? ''
+      }`
```

Comment:
I think `queryToInject` should probably override the existing value in `search` if it exists. And the checking should be `!= null` instead of `=== ''`?

Not a big deal currently as we rarely inject duplicated queries except in the case you showed here, but maybe it's nicer to make this right.

---

Are you also able to add tests [here](https://github.com/vitejs/vite/blob/881d080c293b642685b2448088422102f24821e7/packages/vite/src/node/__tests__/utils.spec.ts#L36)?

---

I think overall about the `?import` issue, it would also be fixed if https://github.com/vitejs/vite/pull/3828 goes through.